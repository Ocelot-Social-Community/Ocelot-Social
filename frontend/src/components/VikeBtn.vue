<template>
  <v-btn
    :variant="isRouteSelected($attrs.href as string) ? 'tonal' : 'flat'"
    @click.prevent="onClick($attrs.href as string)"
  >
    <slot />
  </v-btn>
</template>
<script lang="ts" setup>
import { navigate } from 'vike/client/router'

import { usePageContext } from '#context/usePageContext'

const pageContext = usePageContext()

function onClick(href: string) {
  return navigate(href)
}

const isRouteSelected = (href: string) => {
  if (href === '/app') {
    return pageContext.urlPathname.indexOf(href) === 0
  }
  return pageContext.urlPathname === href
}
</script>
