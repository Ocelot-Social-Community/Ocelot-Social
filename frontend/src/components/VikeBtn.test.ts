import { mount } from '@vue/test-utils'
import { navigate } from 'vike/client/router'
import { describe, it, expect, beforeEach, vi } from 'vitest'

import VikeBtn from './VikeBtn.vue'

vi.mock('vike/client/router')
vi.mocked(navigate).mockResolvedValue()

describe('VikeBtn', () => {
  const Wrapper = () => {
    return mount(VikeBtn, {
      attrs: { href: '/some-path' },
    })
  }
  let wrapper: ReturnType<typeof Wrapper>

  beforeEach(() => {
    wrapper = Wrapper()
  })

  it('renders', () => {
    expect(wrapper.element).toMatchSnapshot()
  })

  it('icon is hidden', () => {
    expect(wrapper.find('.v-icon').exists()).toBe(false)
  })

  describe('with href attribute app', () => {
    beforeEach(async () => {
      await wrapper.setProps({ href: '/app' } as Partial<object>)
    })

    it('has flat variant', () => {
      expect(wrapper.classes()).toContain('v-btn--variant-flat')
    })
  })

  describe('with same href attribute', () => {
    beforeEach(async () => {
      await wrapper.setProps({ href: '/some-url' } as Partial<object>)
    })

    it('has tonal variant', () => {
      expect(wrapper.classes()).toContain('v-btn--variant-tonal')
    })
  })

  describe('click on button', () => {
    it('calls navigate method with given href', async () => {
      await wrapper.find('.v-btn').trigger('click')
      expect(navigate).toHaveBeenCalledWith('/some-path')
    })
  })
})
