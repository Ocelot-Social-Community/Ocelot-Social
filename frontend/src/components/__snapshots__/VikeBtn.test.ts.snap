// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`VikeBtn > renders 1`] = `
<a
  class="v-btn v-theme--light v-btn--density-default v-btn--size-default v-btn--variant-flat"
  href="/some-path"
>
  
  <span
    class="v-btn__overlay"
  />
  <span
    class="v-btn__underlay"
  />
  
  <!---->
  <span
    class="v-btn__content"
    data-no-activator=""
  >
    
    
    
    
    
    
  </span>
  <!---->
  <!---->
</a>
`;
